test/framework/src/main/java/org/elasticsearch/transport/nio/MockNioTransport.java

```diff
@@ -0,0 +1,272 @@
+        MockSocketChannel channel = nioGroup.openChannel(node.getAddress().address(), clientChannelFactory);
+        channel.addConnectListener(ActionListener.toBiConsumer(connectListener));
+        return channel;
+    }
+
+    @Override
+    protected void doStart() {
+        boolean success = false;
+        try {
+            int acceptorCount = 0;
+            boolean useNetworkServer = NetworkService.NETWORK_SERVER.get(settings);
+            if (useNetworkServer) {
+                acceptorCount = 1;
+            }
+            nioGroup = new NioGroup(logger, daemonThreadFactory(this.settings, TRANSPORT_ACCEPTOR_THREAD_NAME_PREFIX), acceptorCount,
+                AcceptorEventHandler::new, daemonThreadFactory(this.settings, TRANSPORT_WORKER_THREAD_NAME_PREFIX),
+                2, TestingSocketEventHandler::new);
+
+            ProfileSettings clientProfileSettings = new ProfileSettings(settings, "default");
+            clientChannelFactory = new MockTcpChannelFactory(clientProfileSettings, "client");
+
+            if (useNetworkServer) {
+                // loop through all profiles and start them up, special handling for default one
+                for (ProfileSettings profileSettings : profileSettings) {
+                    String profileName = profileSettings.profileName;
+                    MockTcpChannelFactory factory = new MockTcpChannelFactory(profileSettings, profileName);
+                    profileToChannelFactory.putIfAbsent(profileName, factory);
+                    bindServer(profileSettings);
+                }
+            }
+
+            super.doStart();
+            success = true;
+        } catch (IOException e) {
+            throw new ElasticsearchException(e);
+        } finally {
+            if (success == false) {
+                doStop();
+            }
+        }
+    }
+
+    @Override
+    protected void stopInternal() {
+        try {
+            nioGroup.close();
+        } catch (Exception e) {
+            logger.warn("unexpected exception while stopping nio group", e);
+        }
+        profileToChannelFactory.clear();
+    }
+
+    private void exceptionCaught(NioSocketChannel channel, Exception exception) {
+        onException((TcpChannel) channel, exception);
+    }
+
+    private void acceptChannel(NioSocketChannel channel) {
+        serverAcceptedChannel((TcpChannel) channel);
+    }
+
+    private static BytesReference toBytesReference(InboundChannelBuffer channelBuffer) {
+        ByteBuffer[] writtenToBuffers = channelBuffer.sliceBuffersTo(channelBuffer.getIndex());
+        ByteBufferReference[] references = new ByteBufferReference[writtenToBuffers.length];
+        for (int i = 0; i < references.length; ++i) {
+            references[i] = new ByteBufferReference(writtenToBuffers[i]);
+        }
+
+        return new CompositeBytesReference(references);
+    }
+
+    private class MockTcpChannelFactory extends ChannelFactory<MockServerChannel, MockSocketChannel> {
```

Comment:
I wonder if we can extract this into a common place. maybe we just have utility methods there that we call. but I would love to reduce the footprint of this class here. I am ok with delegating but it should be mostly shared. We might be able to pass in factories for MockSocketChannel?